src/commands/cmd_profile.c

```diff
@@ -0,0 +1,138 @@
+            CommandCtx_ThreadSafeContextUnlock(qctx);
+            RedisModule_ReplyWithError(ctx, "Graph name already in use as a Redis key.");
+            goto cleanup;
+        }
+        // TODO: free graph if no entities were created.
+    }
+
+    CommandCtx_ThreadSafeContextUnlock(qctx);
+
+    // Perform query validations before and after ModifyAST
+    if (AST_PerformValidations(ctx, ast) != AST_VALID) goto cleanup;
+    ModifyAST(ast);
+    if (AST_PerformValidations(ctx, ast) != AST_VALID) goto cleanup;
+
+    // Acquire the appropriate lock.
+    if(readonly) Graph_AcquireReadLock(gc->g);
+    else Graph_WriterEnter(gc->g);  // Single writer.
+    lockAcquired = true;
+
+    if (ast[0]->indexNode) { // Index operation.
+        RedisModule_ReplyWithError(ctx, "Can't profile index operations.");
+        goto cleanup;
+    }
+
+    resultSet = _prepare_resultset(ctx, ast);
+    ExecutionPlan *plan = NewExecutionPlan(ctx, ast, resultSet, false);
+    ExecutionPlan_Profile(plan);
+    ExecutionPlan_Print(plan, ctx);
+    ExecutionPlanFree(plan);
+
+cleanup:
+    // Release the read-write lock
+    if(lockAcquired) {
+        if(readonly)Graph_ReleaseLock(gc->g);
+        else Graph_WriterLeave(gc->g);
+    }
+
+    ResultSet_Free(resultSet);
+    CommandCtx_Free(qctx);
+}
+
+/* Profiles query
+ * Args:
+ * argv[1] graph name
+ * argv[2] query to profile */
+int MGraph_Profile(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    if (argc < 3) return RedisModule_WrongArity(ctx);
+
+    // Parse AST.
+    char *errMsg = NULL;    
+    const char *query = RedisModule_StringPtrLen(argv[2], NULL);
+
+    AST **ast = ParseQuery(query, strlen(query), &errMsg);
+    if (!ast) {
+        RedisModule_Log(ctx, "debug", "Error parsing query: %s", errMsg);
+        RedisModule_ReplyWithError(ctx, errMsg);
+        free(errMsg);
+        return REDISMODULE_OK;
+    }
+    if(AST_Empty(ast[0])) {
+        AST_Free(ast);
+        RedisModule_ReplyWithError(ctx, "Error empty query.");
+        return REDISMODULE_OK;
+    }
+
+    bool readonly = AST_ReadOnly(ast);
+
+    /* Determin query execution context
+     * queries issued within a LUA script or multi exec block must
+     * run on Redis main thread, others can run on different threads. */
+    CommandCtx *context;
+    int flags = RedisModule_GetContextFlags(ctx);
+    if (flags & (REDISMODULE_CTX_FLAGS_MULTI | REDISMODULE_CTX_FLAGS_LUA)) {
+      // Run query on Redis main thread.
+      context = CommandCtx_New(ctx, NULL, ast, argv[1], argv, argc);
+      _MGraph_Profile(context);
+    } else {
+      // Run query on a dedicated thread.
+      RedisModuleBlockedClient *bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
+      context = CommandCtx_New(NULL, bc, ast, argv[1], argv, argc);
+      thpool_add_work(_thpool, _MGraph_Profile, context);
+    }
+
+    // Replicate only if query has potential to modify key space.
+    if(!readonly) RedisModule_ReplicateVerbatim(ctx);
```

Comment:
I would imagine that like EXPLAIN, PROFILE directives will always be read-only and cannot modify the keyspace.

Reply:
Nope, profile actually runs the query.